letta/server/server.py

```diff
@@ -1758,6 +1760,110 @@ def get_agent_block_by_label(self, user_id: str, agent_id: str, label: str) -> B
                 return block
         return None
 
+    def run_tool(self, tool_id: str, tool_args: str, user_id: str) -> FunctionReturn:
+        """Run a tool using the sandbox and return the result"""
+
+        try:
+            tool_args_dict = json.loads(tool_args)
+        except json.JSONDecodeError:
+            raise ValueError("Invalid JSON string for tool_args")
+
+        # Get the tool by ID
+        user = self.user_manager.get_user_by_id(user_id=user_id)
+        tool = self.tool_manager.get_tool_by_id(tool_id=tool_id, actor=user)
+        if tool.name is None:
+            raise ValueError(f"Tool with id {tool_id} does not have a name")
+
+        # TODO eventually allow using agent state in tools
+        agent_state = None
+
+        try:
+            sandbox_run_result = ToolExecutionSandbox(tool.name, tool_args_dict, user_id).run(agent_state=agent_state)
```

Comment:
a bit of a bit, but can't this be the same code as `run_tool_from_source` once the `Tool` object is ready?

Reply:
oh yeah this code I kind of stopped editing in the PR since I wasn't sure if we wanted to delete it?

e.g. if we're not going to use `run_tool` and only use `run_tool_from_source`, shouldn't we just add the later and kill the former (and add it if we need it later?)